core/lib/tasks/payment_method.rake

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+namespace :payment_method do
+  desc "Deprecates old payment methods and fixes ActiveRecord::SubclassNotFound error, "\
+  "which happens after switching Payment Service Provider."
+  task deprecate_unsupported_payment_methods: :environment do
+    unsupported_ids = get_unsupported_payment_method_ids
+
+    set_unsupported_payment_methods_type(unsupported_ids)
+    # Let ActiveRecord handle boolean format, which may vary
+    # depending on the DBMS in use.
+    set_unsupported_payment_methods_state(unsupported_ids)
+  end
+
+  def set_unsupported_payment_methods_type(unsupported_ids)
+    where_clause = get_where_clause('id', unsupported_ids)
+    ActiveRecord::Base.connection.execute(
+      "UPDATE spree_payment_methods "\
+      "SET type='#{Spree::PaymentMethod.name}' "\
+      "WHERE #{where_clause};"
+    )
+  end
+
+  def set_unsupported_payment_methods_state(unsupported_ids)
+    Spree::PaymentMethod
+      .where(id: unsupported_ids)
+      .update(deprecated: true,
+              active: false,
```

Comment:
Is there a reason why we can't set the type here as well?

Reply:
That's the point of this fix actually: even running a query `Spree::PaymentMethod.where(...)` before changing the `type` via raw sql would make AR immediately raise the STI error, while creating the result instances, so `type` needs to be changed before that.
So, I could have updated the record fully in the first raw sql update, but I preferred to separate this second update to let AR handle boolean values via its connection adapter.

Reply:
Sure, you are right. I'd try to make a single query though, maybe we can move these attributes settings to the raw SQL update?

Reply:
Updating in a single query (and every single row individually, as you suggested below), would make the code something like this.
```ruby
def update_row(id)
    connection = ActiveRecord::Base.connection
    false_value = connection.adapter_name[/sqlite/i] ? '0' : 'false'
    true_value = connection.adapter_name[/sqlite/i] ? '1' : 'true'
    connection.exec_query(
      "UPDATE spree_payment_methods "\
      "SET type='#{Spree::PaymentMethod.name}', deprecated=#{true_value}, "\
      "active=#{false_value}, available_to_users=#{false_value}, available_to_admin=#{false_value} "\
      "WHERE id=#{id}"
    )
end
```
The task becomes really shorter, I like it, but I don't find really nice to manually choose what to write depending on the db in use (it's a db adapter's job), since doing it via AR would be more stable (for example, in case another db will be supported in the future that another "true" value).
Do you think I'm just worrying too much? It's unlikely to support another db that will require another boolean value different from "0"/"1" and "false"/"true", isn't it?

Reply:
ActiveRecord provides facilities for avoiding hardcoding those values, you can use `ActiveRecord::Base.connection.quoted_true` and `ActiveRecord::Base.connection.quoted_false`, this should make the SQL generic enough IMO.

Reply:
Thanks! I couldn't find it myself, this helped me changing the script as suggested by @kennyadsl, I just force pushed the change.